pkg/server/plugin/zmq/broker.go

```diff
@@ -221,50 +243,93 @@ func (lb *Broker) Channeler() {
 			lb.logger.Debugf("zmq/broker: zmq => channel %#x, %s\n", frames[0], frames)
 			// Dispacth back to the channel based on the zmq first frame
 			address := string(frames[0])
-			respChan, ok := lb.addressChan[address]
-			if !ok {
-				lb.logger.Infof("zmq/broker: chan not found for worker %s\n", address)
+			messageType := string(frames[2])
+			bounceCount, err := strconv.Atoi(string(frames[3]))
+			if err != nil {
+				lb.logger.Infof("zmq/broker: Cannot parse bounce_count int %v\n", frames)
 				break
 			}
-			delete(lb.addressChan, address)
-			respChan <- frames[2]
+			requestID := string(frames[4])
+			message := frames[6]
+			if messageType == Request {
+				parcel := newParcel(message)
+				parcel.Worker = address
+				lb.ReqChan <- parcel
+				go func (p *Parcel) {
+					response := <- p.RespChan
+					frames := [][]byte{
+						[]byte(address),
+						[]byte(address),
+						[]byte{},
+						[]byte(Response),
+						[]byte(strconv.Itoa(bounceCount)),
+						[]byte(requestID),
+						[]byte{},
+						response,
+					}
+					lb.respChan <- frames
+				}(parcel)
+			} else if messageType == Response {
+				respChan, ok := lb.addressChan[requestID]
+				if !ok {
+					lb.logger.Infof("zmq/broker: chan not found for worker %s\n", address)
+					break
+				}
+				delete(lb.addressChan, requestID)
+				respChan <- message
+			}
 		case p := <-lb.recvChan:
 			// Save the chan and dispatch the message to zmq
 			// If current no worker ready, will retry after HeartbeatInterval.
 			// Retry for HeartbeatLiveness times
-			address := lb.workers.Next()
+			// TODO(b123400): instead of asking for a new worker
+			// get the current worker (if possible, e.g. when nested, we already have the worker id)
+			// Or, we can put the worker back to pool when received request from plugin? not very good?
+			var address string
+			if p.Worker != "" {
+				address = p.Worker
+			} else {
+				address = lb.workers.Next()
+			}
 			if address == "" {
 				if p.retry < HeartbeatLiveness {
 					p.retry += 1
 					lb.logger.Infof("zmq/broker: no worker available, retry %d...\n", p.retry)
-					go func(p2 *parcel) {
+					go func(p2 *Parcel) {
 						time.Sleep(HeartbeatInterval)
 						lb.recvChan <- p2
 					}(p)
 					break
 				}
 				lb.logger.Infof("zmq/broker: no worker available, timeout.\n")
-				p.respChan <- []byte{0}
+				p.RespChan <- []byte{0}
 				break
 			}
 			addr := []byte(address)
+			requestID := generateRequestID()
 			frames := [][]byte{
 				addr,
 				addr,
 				[]byte{},
-				p.frame,
+				[]byte(Request),
+				[]byte("0"),
+				[]byte(requestID),
+				[]byte{},
+				p.Frame,
 			}
-			lb.addressChan[address] = p.respChan
+			lb.addressChan[requestID] = p.RespChan
```

Comment:
OK, lets put the conversation about request-id in this thread.
Let me coin 2 terms first,

1. __The big request__ - Outer most request, requests that our current implementation can make
2. __The small request__ - Nested request, requests that are made when handling another request.

What you are saying is, request-id should have 2 parts. `PREFIX-SUFFIX`, the prefix refers to the __big__ request, and suffix refers to __small__ requests, right?

For example:

```
@op('foo:hello')
def english():
    send_action('foo:ciao')
    return {'key':'thanks'}

@op('foo:ciao')
def italian():
    return {'key':'grazie'}
```

Would results in:
1. Browser calls /foo/bar
2. Server calls plugin with request-id "ZZZZ-AAAA" (foo:hello)
3. Plugin calls server with request-id "ZZZZ-BBBB" (foo:ciao)
4. Server calls plugin with request-id "ZZZZ-CCCC" (foo:ciao)
5. Plugin responses server with request-id "ZZZZ-CCCC" (foo:ciao)
6. Server responses plugin with request-id "ZZZZ-BBBB" (foo:ciao)
7. Plugin responses server with request-id "ZZZZ-AAAA" (foo:hello)

If the this is wrong, can you provide the correct data for the above code?

Assuming I am correct, saving prefix is not good enough.
1. In step (2), server will save channel for "ZZZZ-AAAA" with the key "ZZZZ".
2. In step (4), server will save channel for "ZZZZ-CCCC" with the key "ZZZZ".
Saving only the suffix is acceptable, but I cannot find any reason not to save the entire request id, after all, `addressChan` is broker level, not request level.